File: src/initial_translation/GRE_stepwise.py
# Python modules
import jax.numpy as jnp
from tqdm import tqdm  # For progress bars

# Application modules
from kernel import kernel
from helper_functions import stepwise
from SPRE_opt import SPRE_opt
from SPRE import SPRE

def GRE_stepwise(X, Y, k_name):
    """
    Stepwise model selection for GRE (Gauss-Richardson Extrapolation).

    Parameters:
        X : jnp.ndarray of shape (n_train, d)
            Training inputs.
        Y : jnp.ndarray of shape (n_train,)
            Training outputs.
        k_name : str
            Kernel name: "Gaussian", "GaussianARD", "Matern1/2", "Matern3/2", or "white".

    Returns:
        out : dict
            Dictionary with predictive mean, variance, and fitted model details.
    """
    d = X.shape[1]  # data dimension

    A = jnp.array([[0, 0]])  # mean function: just the intercept
    B = jnp.zeros((1, d))    # initial rate function: only intercept
    order = 0

    fit = SPRE_opt(A, X, Y, (B, k_name))
    cv = fit["cv"]

    carry_on = True
    while carry_on:
        order += 1
        B_extra = stepwise(B, order)  # Generate all predictors of the next order
        n_extra = B_extra.shape[0]

        print(f"Fitting interactions of order {order}...")

        to_include = jnp.zeros(n_extra, dtype=bool)
        for i in range(n_extra):
            B_new = jnp.vstack([B, B_extra[i, :]])
            fit_new = SPRE_opt(A, X, Y, (B_new, k_name))
            cv_new = fit_new["cv"]

            if cv_new < cv:
                to_include[i] = True

            # Progress bar substitute
            #cwbar((i + 1) / n_extra)

        if jnp.any(to_include):
            B_updated = jnp.vstack([B, B_extra[to_include, :]])
            fit_updated = SPRE_opt(A, X, Y, (B_updated, k_name))
            cv_updated = fit_updated["cv"]

            if cv_updated >= cv:
                carry_on = False
            else:
                fit = fit_updated
                B = B_updated
                cv = cv_updated
        else:
            carry_on = False

        #cwbar("done")

    # Final GP fit with optimal parameters
    x_opt = fit["x"]
    out = SPRE(A, X, Y, x_opt, (B, k_name))

    return out

File: matlab_code/GRE_stepwise.m
%% Stepwise model selection for GRE
%
% Inputs:
% X         = n_train x d, training inputs
% Y         = n_train x 1, training outputs
% k_name    = string, "Gaussian"
%                     "GaussianARD"
%                     "Matern1/2"
%                     "Matern3/2" 
%                  or "white" 
% max_order = integer, maximum polynomial order
%
% Outputs:
%
% out.mu      = scalar, predictive mean for f(0)
% out.cov     = scalar, predictive variance for f(0)
% out.mu_GP   = function R^d -> R, predictive mean for fitted GP
% out.cov_GP  = function R^d x R^d -> R, predictive covariance for fitted GP
% out.mu_cv   = n_train x 1, LOOCV predictive means
% out.var_cv  = n_train x 1, LOOCV predictive variances
% out.cv      = scalar, LOOCV criterion
% out.cv_grad = p x 1, gradient of LOOCV criterion

function out = GRE_stepwise(X,Y,k_name,max_order)  %%%%% NEW

addpath(genpath("helper_functions"))

d = size(X,2); % data dimension
 
A = zeros(1,d); % mean function is just the intercept

B = zeros(1,d); % initialise rate function with just an intercept
order = 0; 
fit = SPRE_opt(A,X,Y,{B,k_name});
cv = fit.cv;

carry_on = true;
while carry_on && (order < max_order)   %%%%%%%%%%%%% NEW
    order = order + 1; % consider the addition of higher-order interations
    B_extra = stepwise(B,order); % all predictors of the next order to consider
    n_extra = size(B_extra,1);

    % monitoring
    cwbar("Fitting interactions of order " + num2str(order,'%u') + ": ")

    to_include = false(n_extra,1);
    for i = 1:n_extra
        B_new = [B; B_extra(i,:)];
        fit_new = SPRE_opt(A,X,Y,{B_new,k_name});
        cv_new = fit_new.cv;
        if cv_new < cv % if adding new predictor helped
            to_include(i) = true;
        end
        cwbar(i/n_extra)
    end
    if sum(to_include) > 0
        B_updated = [B; B_extra(to_include,:)];
        fit_updated = SPRE_opt(A,X,Y,{B_updated,k_name});
        cv_updated = fit_updated.cv; % calculate cv loss for updated model
        if cv_updated >= cv
            carry_on = false; % terminate
        else 
            fit = fit_updated;
            B = B_updated;
            cv = cv_updated;
        end
    else
        carry_on = false; % terminate
    end

    cwbar("!")

end

% optimal parameters
x_opt = fit.x;

% fit GP with optimal parameters
out = SPRE(A,X,Y,x_opt,{B,k_name},false);

end